kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/evm-int-simplification.k

```diff
@@ -57,31 +50,94 @@ module EVM-INT-SIMPLIFICATION-COMMON
   // #asWord
   // ###########################################################################
 
-  rule [asWord-lb]: 0 <=Int #asWord(_WS)     => true [simplification, smt-lemma]
-  rule [asWord-ub]: #asWord(_WS) <Int pow256 => true [simplification, smt-lemma]
-
-  rule [asWord-lt]:
-    #asWord ( BA ) <Int X => true
-    requires 2 ^Int ( 8 *Int lengthBytes(BA) ) <=Int X
-    [concrete(X), simplification]
-
-  rule [asWord-lt-concat]:
-    #asWord ( BA1 +Bytes BA2 ) <Int X => #asWord ( BA1 ) <Int ( X -Int #asWord ( BA2 ) ) /Int ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) )
-    requires ( X -Int #asWord ( BA2 ) ) modInt ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) ) ==Int 0
-    [concrete(BA2, X), simplification, preserves-definedness]
-
-  rule [asWord-lt-range]:
-    #asWord ( #range ( _:Bytes, S, W ) ) <Int X => true
-    requires 0 <=Int S andBool 0 <=Int W
-     andBool 2 ^Int ( 8 *Int W ) <=Int X
-    [simplification, concrete(S, W, X), preserves-definedness]
-
-  rule [lt-asWord-range]:
-    X <Int #asWord ( #range ( B:Bytes, S, W ) ) => X <Int #asWord ( B )
+  // Unification: two `#asWord`s
+  rule [asWord-eq-asWord]:
+    #asWord ( B1 ) ==Int #asWord ( B2 ) => #buf ( 32 -Int lengthBytes(B1), 0 ) +Bytes B1 ==K #buf ( 32 -Int lengthBytes(B2), 0 ) +Bytes B2
+    requires lengthBytes(B1) <=Int 32 andBool lengthBytes(B2) <=Int 32
+    [simplification, preserves-definedness]
+
+  // Unification: two `#asWord`s
+  rule [asWord-eq-asWord-ml]:
+    { #asWord ( B1 ) #Equals #asWord ( B2 ) } => { true #Equals #asWord ( B1 ) ==Int #asWord ( B2 ) }
+    [simplification, preserves-definedness]
+
+  // Unification: `#asWord` and concrete number
+  rule [asWord-eq-num]:
+    #asWord ( B:Bytes ) ==Int X:Int =>
+      0 <=Int X andBool X <Int 2 ^Int (8 *Int minInt(32, lengthBytes(B)))
+                andBool (
+                  ( lengthBytes(B) <=Int 32 andBool #buf ( 32 -Int lengthBytes(B), 0 ) +Bytes B ==K #buf ( 32, X ) )
+                    orBool
+                  ( lengthBytes(B) >Int 32 andBool #range ( B, lengthBytes(B) -Int 32, 32 ) ==K #buf ( 32, X ) )
+                )
+    [simplification, concrete(X), comm, preserves-definedness]
```

Comment:
I think we can remove the second case (`lengthBytes(B) >Int 32`) entirely, because the rule below `asWord-trim-overflowing` will apply in that case. So we'll end up with:

```
#asWord(#range(B, lengthBytes(B) -Int 32, 32)) ==Int X
```

Which will fall into the first case.

```suggestion
  // Unification: `#asWord` and concrete number
  rule [asWord-eq-num]:
    #asWord ( B:Bytes ) ==Int X:Int => 0 <=Int X andBool X <Int 2 ^Int (8 *Int minInt(32, lengthBytes(B)))
                               andBool #buf ( 32 -Int lengthBytes(B), 0 ) +Bytes B ==K #buf ( 32, X )
    requires lengthBytes(B) <=Int 32
    [simplification, concrete(X), comm, preserves-definedness]
```
```

Reply:
I think so, yes, but only if we can say with certainty whether the length of the buffer `B` is `<=Int` or `>Int` than 32, right? 

While I think that this should always be the case when `#asWord` is used, I don't think that we have a 100% guarantee that `#asWord` will not get a parameter whose length is sufficiently symbolic for us not to be able to tell that, in which case `asWord-trim-overflowing` won't fire and we we need to have both cases?

Reply:
I guess, do we have an example of the case you're describing from client engagements? If so, definitely add a test for it. If not, I'm somewhat erring on the side of the simpler more auditable lemma.

Reply:
I don't have an example from client engagements and I believe that the case I'm describing won't happen. What can happen is that the slot update mechanism adds leading zeros that take the buffer length beyond 32, but in those cases we would know the length precisely.

I'm happy to simplify the lemma.

Reply:
Actually, can we do this, even simpler, we don't need either the minimum or the leading zeros?
```
  rule [asWord-eq-num]:
    #asWord ( B:Bytes ) ==Int X:Int =>
      0 <=Int X andBool X <Int 2 ^Int (8 *Int lengthBytes(B))
      andBool B ==K #buf ( lengthBytes(B), X )
    requires lengthBytes(B) <=Int 32
    [simplification, concrete(X), comm, preserves-definedness]
```

Reply:
If we're simplifying, we can probably go even further, and have an extra `requires` clause `X <Int pow256`. We can have a separate case for otherwise:

```
rule #asWord(_) ==Int X => false requires X <Int 0 orBool pow256 <=Int X [simplification(40)]
```

Reply:
That would allow us to instead have:

```suggestion
  // Unification: `#asWord` and concrete number
  rule [asWord-eq-num]:
    #asWord ( B:Bytes ) ==Int X:Int => #buf ( 32 -Int lengthBytes(B), 0 ) +Bytes B ==K #buf ( 32, X )
    requires lengthBytes(B) <=Int 32 andBool #rangeUInt(256, X)
    [simplification, concrete(X), comm, preserves-definedness]

  rule [asWord-neq-num]:
    #asWord(_) ==Int X => false requires notBool #rangeUInt(256, X) [simplification]
```

That actually is slightly more general since the second one is not restricting us to `concrete(X)`.

Reply:
The first one is `concrete(X)` anyway, so I think that this shouldn't really be any less applicable.

Reply:
and we will always end up only with `B ==K #buf ( lengthBytes(B), X )` really.

Reply:
The second simplification you suggested, `[asWord-neq-num]`, if `X` is not concrete, I think would fire always for any `X` of arbitrary complexity, which means many calls to the SMT solver potentially about the range, and it almost always not applying.